rust/chains/hyperlane-cosmos/src/providers/rpc.rs

```diff
@@ -190,47 +217,25 @@ impl WasmIndexer for CosmosWasmIndexer {
     }
 
     #[instrument(err, skip(self, parser))]
-    async fn get_range_event_logs<T>(
+    async fn get_event_log<T>(
         &self,
-        range: RangeInclusive<u32>,
+        block_number: u32,
         parser: for<'a> fn(&'a Vec<EventAttribute>) -> ChainResult<ParsedEvent<T>>,
     ) -> ChainResult<Vec<(T, LogMeta)>>
     where
-        T: PartialEq + Send + Sync + 'static,
+        T: Send + Sync + PartialEq + 'static,
     {
-        // Page starts from 1
-        let query = Query::default()
-            .and_gte("tx.height", *range.start() as u64)
-            .and_lte("tx.height", *range.end() as u64)
-            .and_eq(
-                format!("{}._contract_address", self.target_event_kind),
-                self.contract_address.address(),
-            );
-
-        let tx_search_result = self.tx_search(query.clone(), 1).await?;
-
-        // Using the first tx_search_result, we can calculate the total number of pages.
-        let total_count = tx_search_result.total_count;
-        let last_page = div_ceil(total_count, PAGINATION_LIMIT.into());
+        let client = self.provider.rpc().clone();
+        let block_fut = tokio::spawn(async move { client.block(block_number).await });
 
-        let mut logs = self
-            .handle_txs(tx_search_result.txs, parser)?
-            .collect::<Vec<_>>();
+        let client = self.provider.rpc().clone();
+        let block_results_fut =
+            tokio::spawn(async move { client.block_results(block_number).await });
 
-        // If there are any more pages, fetch them and append to the result.
-        for page in 2..=last_page {
-            trace!(page, "Performing tx search");
+        let (block_res, block_results_res) = tokio::join!(block_fut, block_results_fut);
```

Comment:
for my own learning, is there any difference between this approach of spawning 2 tasks and joining vs this? does the following just spawn tasks under the hood?
```
let client1 = self.provider.rpc().clone();
let client2 = self.provider.rpc().clone();
let (block_res, block_results_res) = tokio::join!(
    client1.block(block_number),
    client2.block_results(block_number),
);
```

Reply:
simplified - didn't know it could be done this way. It just [polls](https://docs.rs/tokio/latest/src/tokio/macros/join.rs.html#126) every item in the `join!` as a future, until completion. There is a [wrapper](https://docs.rs/tokio/latest/src/tokio/future/maybe_done.rs.html#13) around the items but it's very light